---
title: no-constant-condition-with-break
description: Disallow constant conditions, but allow constant loops with break, return, or configured exit calls.
---

import { Aside, Code, LinkCard, CardGrid, Card, Tabs, TabItem } from "@astrojs/starlight/components";

Disallow constant conditions, but allow constant loops that include exits like `break`, `return`, or configured calls.

## Rule Details

This rule is similar to ESLint's built-in `no-constant-condition` rule, but it makes an exception for constant loops
(like `while (true)`) that contain an exit path (for example `break` or `return`). You can also configure additional
call expressions (for example `coroutine.yield`) as loop exits via `loopExitCalls`.

<Aside type="tip" title="When to use this rule">
	Use this rule when you want to catch accidental constant conditions in `if` statements and loops without a loop exit,
	while allowing intentional infinite loops with proper exit mechanisms.
</Aside>

## Options

```typescript
{
	"cease-nonsense/no-constant-condition-with-break": [
		"error",
		{
			loopExitCalls: ["coroutine.yield", "task.wait"]
		}
	]
}
```

- `loopExitCalls` (`string[]`) - Fully-qualified call paths treated as loop exits inside constant loops.

## Examples

<Tabs>
	<TabItem label="❌ Incorrect" icon="close">
		### Constant Conditions Without Loop Exit
		These patterns are flagged because they are likely bugs or missing exit conditions.

    	<Code
    		code={[
    			"// if statement with constant condition",
    			'if (true) { doSomething(); }',
    			'',
			"// while loop without a loop exit",
			'while (true) { doSomething(); }',
    			'',
			"// for loop with constant condition and no loop exit",
			'for (;; true) { doSomething(); }',
    			'',
			"// do-while loop without a loop exit",
			'do { doSomething(); } while (true);',
    		].join("\n")}
    		lang="tsx"
    	/>
    </TabItem>
    <TabItem label="✅ Correct" icon="check">
    	### Valid Patterns
		These patterns are allowed because they either have proper exit conditions or intentional loop exits.

    	<Code
    		code={[
    			"// Variable condition is fine",
    			'while (condition) { doSomething(); }',
    			'',
			"// Loop with break is allowed",
			'while (true) {',
			'\tif (shouldStop) break;',
			'\tdoSomething();',
			'}',
			'',
			"// Loop with return is allowed",
			'function run() {',
			'\twhile (true) {',
			'\t\tif (shouldStop) return;',
			'\t\tdoSomething();',
			'\t}',
			'}',
			'',
			"// Configured exit call (with loopExitCalls)",
			'while (true) {',
			'\tcoroutine.yield();',
			'\tdoSomething();',
			'}',
			'',
			"// Labeled break targeting this loop",
			'outer: while (true) {',
			'\tif (done) break outer;',
			'\tdoSomething();',
			'}',
			'',
			"// For loop with break",
			'for (;; true) {',
			'\tif (condition) break;',
			'\tdoSomething();',
			'}',
    		].join("\n")}
    		lang="tsx"
    	/>
    </TabItem>

</Tabs>

## Detection Strategy

<CardGrid stagger>
	<Card title="Constant Evaluation" icon="magnifier">
		Evaluates expressions to determine if they always result in the same value.
	</Card>
	<Card title="Exit Path Analysis" icon="target">
		Traces loop exits (`break`, `return`) and configurable exit calls, and verifies labeled breaks target the same loop.
	</Card>
	<Card title="Nested Loop Aware" icon="seti:folder">
		Correctly handles nested loops and labeled breaks.
	</Card>
	<Card title="Function Boundary" icon="seti:lock">
		Respects function boundaries—breaks inside callbacks don't count as exiting the outer loop.
	</Card>
</CardGrid>

## Related Rules

<CardGrid>
	<LinkCard
		title="no-useless-use-effect"
		description="Prevents unnecessary useEffect patterns"
		href="/eslint-cease-nonsense-rules/rules/no-useless-use-effect/"
	/>
</CardGrid>

## Further Reading

<CardGrid>
	<LinkCard
		title="ESLint no-constant-condition"
		description="Original ESLint rule documentation"
		href="https://eslint.org/docs/latest/rules/no-constant-condition"
	/>
</CardGrid>
