---
title: prevent-abbreviations
---

Prevent abbreviations in variable and property names.

## Rule Details

This rule discourages the use of abbreviations in favor of more descriptive names. It provides suggestions for replacements and can automatically fix single-replacement cases.

## Options

```jsonc
{
	"cease-nonsense/prevent-abbreviations": [
		"error",
		{
			"checkFilenames": true,
			"checkProperties": false,
			"checkVariables": true,
			"replacements": {},
			"allowList": {},
			"ignore": [],
		},
	],
}
```

### Configuration Parameters

- **checkFilenames** (default: `true`): Check file names for abbreviations.
- **checkProperties** (default: `false`): Check object properties and method names.
- **checkVariables** (default: `true`): Check variable names.
- **replacements**: Custom replacement mappings (e.g., `{ "args": { "arguments": true } }`).
- **allowList**: List of abbreviations to allow (e.g., `{ "props": true }`).
- **ignore**: Array of regex patterns or strings to ignore.

## Examples

### ❌ Incorrect

```typescript
const err = new Error();
const fn = () => {};
const dist = calculateDistance();
```

### ✅ Correct

```typescript
const error = new Error();
const func = () => {};
const distance = calculateDistance();
```

## When Not To Use It

If your codebase has established abbreviation conventions (e.g., `req`/`res` for Express handlers) that you want to keep, use the `allowList` option or disable this rule for those files.

## Performance

This rule is optimized with:

- Aggressive caching of word and name replacements
- Pre-compiled regex patterns
- Early exits for constants and allow-listed names

## Related Rules

- `unicorn/prevent-abbreviations` - Original rule (this is a faster replacement)
