---
title: require-react-display-names
---

Require `displayName` property on exported `React.memo` components and `React.createContext` contexts for better debugging.

## Rule Details

React components wrapped in `memo` and contexts created with `createContext` lose their names in React DevTools unless you explicitly set `displayName`. This rule enforces proper naming patterns to maintain debuggability.

## Examples

### ❌ Incorrect

**Direct memo export (worst pattern)**

```typescript
import { memo } from "@rbxts/react";

function ButtonNoMemo() {
	return <button />;
}

// Error: Do not export memo result directly
export default memo(ButtonNoMemo);
```

**Missing displayName on memo**

```typescript
import { memo } from "@rbxts/react";

function ChecklistClaimButtonNoMemo() {
	return <frame />;
}

// Error: Missing displayName
const ChecklistClaimButton = memo(ChecklistClaimButtonNoMemo);
export default ChecklistClaimButton;
```

**Direct createContext export**

```typescript
import React from "@rbxts/react";

// Error: Do not export createContext result directly
export default React.createContext<string | undefined>(undefined);
```

**Missing displayName on context**

```typescript
import React from "@rbxts/react";

// Error: Missing displayName
const ErrorBoundaryContext = React.createContext<unknown>(undefined);
export default ErrorBoundaryContext;
```

### ✅ Correct

**Proper memo pattern**

```typescript
import { memo } from "@rbxts/react";

function ChecklistClaimButtonNoMemo() {
	return <frame />;
}

export const ChecklistClaimButton = memo(ChecklistClaimButtonNoMemo);
ChecklistClaimButton.displayName = "ChecklistClaimButton";
export default ChecklistClaimButton;
```

**Proper context pattern**

```typescript
import React from "@rbxts/react";

const ErrorBoundaryContext = React.createContext<unknown>(undefined);
ErrorBoundaryContext.displayName = "ErrorBoundaryContext";
export default ErrorBoundaryContext;
```

**Internal (non-exported) usage is fine**

```typescript
import { memo, createContext } from "@rbxts/react";

// Not exported, so no displayName required
const InternalContext = createContext<number>(0);

function InnerComponentNoMemo() {
	return <div />;
}

// Not exported, so no displayName required
const InnerComponent = memo(InnerComponentNoMemo);

// Only this is exported
export function Container() {
	return <InnerComponent />;
}
```

## Options

### `environment`

- Type: `"roblox-ts" | "standard"`
- Default: `"roblox-ts"`

Controls which React package sources are checked.

- `"roblox-ts"`: Checks imports from `@rbxts/react` and `@rbxts/roact`
- `"standard"`: Checks imports from `react` and `react-dom`

**Configuration for standard React**

```typescript
{
  "cease-nonsense/require-react-display-names": ["error", {
    "environment": "standard"
  }]
}
```

## Why Is This Bad?

### DevTools Debugging

Without `displayName`, memoized components and contexts appear as anonymous or with unhelpful names in React DevTools:

```text
<Memo(Anonymous)>     // vs <ChecklistClaimButton>
<Context.Provider>    // vs <ErrorBoundaryContext.Provider>
```

### Production Debugging

Minification often strips function names. `displayName` survives minification and appears in error stack traces and profiler data.

### Direct Export Problems

Exporting `memo(Component)` directly:

1. Loses the component name completely
2. Makes it impossible to set `displayName` after the fact
3. Prevents named exports alongside default

## When Not To Use It

- For non-exported memo/context usage (the rule already ignores these)
- If you're using a bundler/transformer that automatically adds displayName
- For very simple internal components where debugging isn't a concern

## Migration Guide

### Before

```typescript
import { memo } from "@rbxts/react";

function ComponentNoMemo(props: Props) {
	return <frame />;
}

export default memo(ComponentNoMemo);
```

### After

```typescript
import { memo } from "@rbxts/react";

function ComponentNoMemo(props: Props) {
	return <frame />;
}

export const Component = memo(ComponentNoMemo);
Component.displayName = "Component";
export default Component;
```

## Related Rules

- [ban-react-fc](./ban-react-fc.md) - Bans React.FC which also causes DevTools issues
